File: team417/src/main/java/org/firstinspires/ftc/team417/roadrunner/HolonomicKinematics.java
package org.firstinspires.ftc.team417.roadrunner;

import static java.lang.Math.abs;
import static java.lang.Math.sqrt;

import androidx.annotation.NonNull;

import com.acmerobotics.roadrunner.Arclength;
import com.acmerobotics.roadrunner.DualNum;
import com.acmerobotics.roadrunner.Pose2d;
import com.acmerobotics.roadrunner.Pose2dDual;
import com.acmerobotics.roadrunner.PosePath;
import com.acmerobotics.roadrunner.PoseVelocity2d;
import com.acmerobotics.roadrunner.PoseVelocity2dDual;
import com.acmerobotics.roadrunner.Twist2dDual;
import com.acmerobotics.roadrunner.Vector2dDual;
import com.acmerobotics.roadrunner.VelConstraint;

import java.util.Arrays;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.OptionalDouble;

public class HolonomicKinematics {
    public KinematicType kinematicType;
    public final double trackWidth;
    public final double lateralMultiplier;

    /**
     * Constructor for HolonomicKinematics.
     *
     * @param kinematicType      Type of kinematic mechanism (Mecanum or X).
     * @param trackWidth         Distance between wheels on opposite sides.
     */
    public HolonomicKinematics(KinematicType kinematicType, double trackWidth) {
        this(kinematicType, trackWidth, 1);
    }

    /**
     * Constructor for HolonomicKinematics.
     *
     * @param trackWidth         Distance between wheels on opposite sides.
     * @param lateralMultiplier  Factor to adjust strafe velocity, ignored for X kinematic type.
     */
    public HolonomicKinematics(double trackWidth, double lateralMultiplier) {
        this.trackWidth = trackWidth;
        this.lateralMultiplier = lateralMultiplier;
    }

    /**
     * Constructor for HolonomicKinematics.
     *
     * @param kinematicType      Type of kinematic mechanism (Mecanum or X).
     * @param trackWidth         Distance between wheels on opposite sides.
     * @param lateralMultiplier  Factor to adjust strafe velocity, ignored for X kinematic type.
     */
    public HolonomicKinematics(KinematicType kinematicType, double trackWidth, double lateralMultiplier) {
        this.kinematicType = kinematicType;
        this.trackWidth = trackWidth;
        this.lateralMultiplier = lateralMultiplier;
    }

    /**
     * Overloaded constructor.
     *
     * @param kinematicType      Type of kinematic mechanism (Mecanum or X).
     * @param trackWidth         Distance between wheels on opposite sides.
     * @param wheelbase          Distance between wheels on the same side.
     * @param lateralMultiplier  Factor to adjust strafe velocity, ignored for X kinematic type.
     */
    public HolonomicKinematics(KinematicType kinematicType, double trackWidth, double wheelbase, double lateralMultiplier) {
        this(kinematicType, (trackWidth + wheelbase) / 2, lateralMultiplier);
    }

    public static class WheelIncrements<Param> {
        public final DualNum<Param> leftFront;
        public final DualNum<Param> leftBack;
        public final DualNum<Param> rightBack;
        public final DualNum<Param> rightFront;

        public WheelIncrements(DualNum<Param> leftFront, DualNum<Param> leftBack, DualNum<Param> rightBack, DualNum<Param> rightFront) {
            this.leftFront = leftFront;
            this.leftBack = leftBack;
            this.rightBack = rightBack;
            this.rightFront = rightFront;
        }
    }

    public <Param> Twist2dDual<Param> forward(WheelIncrements<Param> w) {
        switch (kinematicType) {
            case MECANUM:
                return new Twist2dDual<>(
                        new Vector2dDual<>(
                                (w.leftFront.plus(w.leftBack).plus(w.rightBack).plus(w.rightFront)).times(0.25),
                                (w.leftFront.unaryMinus().plus(w.leftBack).minus(w.rightBack).plus(w.rightFront)).times(0.25 / lateralMultiplier)
                        ),
                        (w.leftFront.unaryMinus().minus(w.leftBack).plus(w.rightBack).plus(w.rightFront)).times(0.25 / trackWidth)
                );
            case X:
                return new Twist2dDual<>(
                        new Vector2dDual<>(
                                (w.leftFront.plus(w.leftBack).plus(w.rightBack).plus(w.rightFront)).times(0.25 / sqrt(2.0)),
                                (w.leftFront.unaryMinus().plus(w.leftBack).minus(w.rightBack).plus(w.rightFront)).times(0.25 / sqrt(2.0))
                        ),
                        (w.leftFront.unaryMinus().minus(w.leftBack).plus(w.rightBack).plus(w.rightFront)).times(0.25 / trackWidth)
                );
            default:
                throw new IllegalStateException("Unexpected kinematicType: " + kinematicType);
        }
    }

    public static class WheelVelocities<Param> {
        public final DualNum<Param> leftFront;
        public final DualNum<Param> leftBack;
        public final DualNum<Param> rightBack;
        public final DualNum<Param> rightFront;

        public WheelVelocities(DualNum<Param> leftFront, DualNum<Param> leftBack, DualNum<Param> rightBack, DualNum<Param> rightFront) {
            this.leftFront = leftFront;
            this.leftBack = leftBack;
            this.rightBack = rightBack;
            this.rightFront = rightFront;
        }

        public List<DualNum<Param>> all() {
            return Arrays.asList(leftFront, leftBack, rightBack, rightFront);
        }
    }

    public <Param> WheelVelocities<Param> inverse(PoseVelocity2dDual<Param> t) {
        switch (kinematicType) {
            case MECANUM:
                return new WheelVelocities<>(
                        t.linearVel.x.minus(t.linearVel.y.times(lateralMultiplier)).minus(t.angVel.times(trackWidth)),
                        t.linearVel.x.plus(t.linearVel.y.times(lateralMultiplier)).minus(t.angVel.times(trackWidth)),
                        t.linearVel.x.minus(t.linearVel.y.times(lateralMultiplier)).plus(t.angVel.times(trackWidth)),
                        t.linearVel.x.plus(t.linearVel.y.times(lateralMultiplier)).plus(t.angVel.times(trackWidth))
                );
            case X:
                return new WheelVelocities<>(
                        t.angVel.times(trackWidth).unaryMinus().times(1/sqrt(2.0)).plus(t.linearVel.x.minus(t.linearVel.y)),
                        t.angVel.times(trackWidth).unaryMinus().times(1/sqrt(2.0)).plus(t.linearVel.x.plus(t.linearVel.y)),
                        t.angVel.times(trackWidth).times(1/sqrt(2.0)).plus(t.linearVel.x.minus(t.linearVel.y)),
                        t.angVel.times(trackWidth).times(1/sqrt(2.0)).plus(t.linearVel.x.plus(t.linearVel.y))


//                        (t.linearVel.x.minus(t.linearVel.y).minus(t.angVel.times(trackWidth))),
//                        (t.linearVel.x.plus(t.linearVel.y).minus(t.angVel.times(trackWidth))),
//                        (t.linearVel.x.minus(t.linearVel.y).plus(t.angVel.times(trackWidth))),
//                        (t.linearVel.x.plus(t.linearVel.y).plus(t.angVel.times(trackWidth)))
                );
            default:
                throw new IllegalStateException("Unexpected kinematicType: " + kinematicType);
        }
    }

    public class WheelVelConstraint implements VelConstraint {
        public final double maxWheelVel;

        public WheelVelConstraint(double maxWheelVel) {
            this.maxWheelVel = maxWheelVel;
        }

        @Override
        public double maxRobotVel(Pose2dDual<Arclength> robotPose, @NonNull PosePath path, double s) {
            Pose2d txRobotWorld = robotPose.value().inverse();
            PoseVelocity2d robotVelWorld = robotPose.velocity().value();
            PoseVelocity2d robotVelRobot = txRobotWorld.times(robotVelWorld);
            OptionalDouble result = inverse(PoseVelocity2dDual.constant(robotVelRobot, 1))
                    .all()
                    .stream()
                    .mapToDouble(wheelVel -> abs(maxWheelVel / wheelVel.value())).min();
            if (result.isPresent()) {
                return result.getAsDouble();
            } else {
                throw new NoSuchElementException("OptionalDouble result had no double present");
            }
        }
    }
}

File: team417/src/main/java/org/firstinspires/ftc/team417/roadrunner/HolonomicKinematics.kt
package org.firstinspires.ftc.team417.roadrunner

import com.acmerobotics.roadrunner.Arclength
import com.acmerobotics.roadrunner.DualNum
import com.acmerobotics.roadrunner.Pose2dDual
import com.acmerobotics.roadrunner.PosePath
import com.acmerobotics.roadrunner.PoseVelocity2dDual
import com.acmerobotics.roadrunner.Twist2dDual
import com.acmerobotics.roadrunner.Vector2dDual
import com.acmerobotics.roadrunner.VelConstraint
import kotlin.math.abs
import kotlin.math.sqrt

enum class KinematicType {
    MECANUM,
    X,
}

/**
 * @param[trackWidth] distance between wheels on opposite sides; see the diagram below
 * ![Wheelbase and track width diagram](https://upload.wikimedia.org/wikipedia/commons/5/52/Wheelbase_and_Track.png)
 * @param[lateralMultiplier] factor that multiplies strafe velocity to compensate for slip; increase it to boost the
 * distance traveled in the strafe direction (ignore this if [kinematicType] is X)
 */
data class HolonomicKinematics @JvmOverloads constructor(
    @JvmField
    val kinematicType: KinematicType,
    @JvmField
    val trackWidth: Double,
    @JvmField
    val lateralMultiplier: Double = 1.0
) {
    /**
     * @param[wheelbase] distance between wheels on the same side; see the diagram in [HolonomicKinematics]
     */
    constructor(
        kinematicType: KinematicType,
        trackWidth: Double,
        wheelbase: Double,
        lateralMultiplier: Double = 1.0
    ) : this(kinematicType, (trackWidth + wheelbase) / 2, lateralMultiplier)

    data class WheelIncrements<Param>(
        @JvmField
        val leftFront: DualNum<Param>,
        @JvmField
        val leftBack: DualNum<Param>,
        @JvmField
        val rightBack: DualNum<Param>,
        @JvmField
        val rightFront: DualNum<Param>,
    )

    fun <Param> forward(w: WheelIncrements<Param>) = Twist2dDual(
        when (kinematicType) {
            KinematicType.MECANUM -> Vector2dDual(
                (w.leftFront + w.leftBack + w.rightBack + w.rightFront) * 0.25,
                (-w.leftFront + w.leftBack - w.rightBack + w.rightFront) * (0.25 / lateralMultiplier),
            )

            KinematicType.X -> Vector2dDual(
                // X-Drive's forward kinematics needs to account for the 45-degree angle of wheels, hence the sqrt(2)/2 multiplier
                (w.leftFront + w.leftBack + w.rightBack + w.rightFront) * (0.25 / sqrt(2.0)),
                (-w.leftFront + w.leftBack - w.rightBack + w.rightFront) * (0.25 / sqrt(2.0))
            )
        },
        (-w.leftFront - w.leftBack + w.rightBack + w.rightFront) * (0.25 / trackWidth),
    )

    data class WheelVelocities<Param>(
        @JvmField
        val leftFront: DualNum<Param>,
        @JvmField
        val leftBack: DualNum<Param>,
        @JvmField
        val rightBack: DualNum<Param>,
        @JvmField
        val rightFront: DualNum<Param>,
    ) {
        fun all() = listOf(leftFront, leftBack, rightBack, rightFront)
    }


    fun <Param> inverse(t: PoseVelocity2dDual<Param>) = when (kinematicType) {
        KinematicType.MECANUM -> WheelVelocities(
            t.linearVel.x - t.linearVel.y * lateralMultiplier - t.angVel * trackWidth,
            t.linearVel.x + t.linearVel.y * lateralMultiplier - t.angVel * trackWidth,
            t.linearVel.x - t.linearVel.y * lateralMultiplier + t.angVel * trackWidth,
            t.linearVel.x + t.linearVel.y * lateralMultiplier + t.angVel * trackWidth,
        )

        KinematicType.X -> WheelVelocities(
            // The inverse kinematics of an X-Drive must consider the 45-degree orientation of the wheels
            (t.linearVel.x - t.linearVel.y - t.angVel * trackWidth) * sqrt(2.0),
            (t.linearVel.x + t.linearVel.y - t.angVel * trackWidth) * sqrt(2.0),
            (t.linearVel.x - t.linearVel.y + t.angVel * trackWidth) * sqrt(2.0),
            (t.linearVel.x + t.linearVel.y + t.angVel * trackWidth) * sqrt(2.0),
        )
    }

    inner class WheelVelConstraint(@JvmField val maxWheelVel: Double) : VelConstraint {
        override fun maxRobotVel(
            robotPose: Pose2dDual<Arclength>,
            path: PosePath,
            s: Double
        ): Double {
            val txRobotWorld = robotPose.value().inverse()
            val robotVelWorld = robotPose.velocity().value()
            val robotVelRobot = txRobotWorld * robotVelWorld
            return inverse(PoseVelocity2dDual.constant<Arclength>(robotVelRobot, 1))
                .all()
                .minOf { abs(maxWheelVel / it.value()) }
        }
    }
}